File: LeeYooseok/BOJ_오늘의 문제/2024/2024-04/2024-04-21/20955 민서의 응급 수술.py
import sys

input = sys.stdin.readline

# union & find 문제

N, M = map(int, input().split())

parent = [i for i in range(N + 1)]


def find(node):
    if parent[node] != node:
        return find(parent[node])
    return parent[node]


def union(node1, node2):
    node1 = find(node1)
    node2 = find(node2)

    if node1 < node2:
        parent[node2] = node1
    else:
        parent[node1] = node2


answer = 0

for _ in range(M):
    u, v = map(int, input().split())
    # union 하기 전에 부모가 같다면, cycle 이 존재하는 것.
    if find(u) == find(v):
        answer += 1  # edge 를 제거 -> 실제로는 제거하지 않은
    else:
        union(u, v)

parent_set = set()
for i in range(1, N + 1):
    parent_set.add(find(i))

print(answer + len(parent_set) - 1)

File: gunkim/2024_Weekly_I_Solved/24_04/20955_민서의응급수술.java
// 순환 그래프일 때 끊는 과정을 처음에 만들지 않았었음
import java.util.*;
import java.io.*;

public class Main {

    private int n;
    private int m;
    private int[] parent;
    private int[] depth;


    public void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        parent = new int[n + 1];
        depth = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            parent[i] = i;
            depth[i] = 1;
        }

        int answer = 0;
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int left = Integer.parseInt(st.nextToken());
            int right = Integer.parseInt(st.nextToken());

            // 이미 같은 트리에 속해있던 거라면 다시 잇는 과정을 하지 말아야 함
            // 즉 끊는 과정으로 카운트 + 1
            if (!union(left, right)) {
                answer++;
            }
        }


        // 노드 1이랑 다 비교
        for (int i = 2; i <= n; i++) {
            if (find(1) != find(i)) {
                union(1, i);
                answer += 1;
            }
        }

        System.out.println(answer);


    }

    private int find(int node) {
        if (parent[node] != node) {
            return parent[node] = find(parent[node]);
        }
        return node;
    }

    private boolean union(int node1, int node2) {
        int root1 = find(node1);
        int root2 = find(node2);

        if (root1 == root2) return false;

        if (depth[root1] > depth[root2]) {
            parent[root2] = root1;
            depth[root2] = depth[root1];
        } else {
            parent[root1] = root2;
            if (depth[root1] == depth[root2]) {
                depth[root2] += 1;
            }
            depth[root1] = depth[root2];
        }
        return true;
    }



    public static void main(String[] args) throws Exception {
        new Main().solution();
    }

}